### DISCUSSION TITLE
Epoch and max train Epoch difference?

### DISCUSSION THREAD

**Discussion by radry** [bmaltais/kohya_ss#1350] [Q&A]
When training Lora there is "Epoch" and "Max train Epoch". The later is clear, it defines how many Epochs are trained but what does "Epoch" do? It doesn't seem to have an influnce on the total steps.

**Comment by vauxra** [comment]
from the docs https://github.com/bmaltais/kohya_ss/wiki/LoRA-training-parameters#epoch
  
Epoch
One epoch is "one set of learning".

For example, let's say you want to learn by reading 50 images each 10 times. In this case, 1 epoch is 50x10 = 500 trainings. If it is 2 epochs, this will be repeated twice, so it will be 500x2 = 1000 times of learning.

After training for the specified number of epochs, a LoRA file will be created and saved to the specified location.

For LoRA, 2-3 epochs of learning is sufficient.

**Comment by 5KilosOfCheese** [comment]
Epoch just basically means a cycle of time. Example year as an epoch is from first calendar day to the last calendar day, then we start a year again. Years can be different amount of time in them, but every year is always one year. One epoch here is just one repeat of the dataset. So if you have 10 images and 10 repeats, one epoch is 10 * 10 = 100 steps. So if you train for 10 epochs, it is 1000 steps.

If you prefer to use something like polynomial, sine, repeating, linear, or other LR-curve, you can fit the curve to the specific range (like 30 epochs). So you can do a short test run of 10 epochs, and the LR curve will remain the same as if you were doing 30. So the max epoch allows you to clamp that fixed, but allows you to to datasets which are different in size and repeats, and training in different amounts of steps or epochs.

This is very relevant when you switch between optimisers, as different optimisers prefer different tactics; this is due to how the math in them is done, some perform best when number of epochs is greater than repeats, while some prefer more repeats than epocs (Adafactor), while some really don't care one way or another because they work based on just steps overall.

I personally only train in full epochs. I really don't care about steps. However others take different approach. Keep in mind that the current state of this stuff is bit... obscure and arcane, there are lots of misinformation, irrelevant information, outright beliefs and old wives tales, going about how to train LoRA's. You need to just practice, do test runs and try out different settings, to find the tactic that produces best results for you.

Because keep in mind that EVERY SINGLE concept you try to do is unique. Every single training needs adjustments in it's settings to get best results. You learn to read the tea leaves and signs after a while, so you'll get a general gut feeling on what to do to get it right. And what is "right" in images is basically entirely subjective.
